### ISSUE TITLE
Include library of standard icons

### ISSUE THREAD

**Issue by buddha314** [beeware/toga#2441]
### What is the problem or limitation you are having?

Libraries like Material Design include several platform specific icons.  Things like un/filled stars for ratings.  Toga only provides the default icon (which is adorable, btw)

### Describe the solution you'd like

The ability to do something like `icon=toga.Icon.star-filled`

### Describe alternatives you've considered

downloading entire libraries to my project folder

### Additional context

briefly discussed on Discord

**Comment by freakboy3742** [comment]
Agreed that this is a gap that Toga should fill.

I can see three approaches that could be taken.

The first is to only wrap the icons for which there is a common platform-specific representation. I've fairly sure all the platforms have "standard" icons for features like "home", "share", "download", "upload", and so on; we would work out what the lowest common subset of those icons are, and provide a guaranteed Toga representation for those icons. The question is whether there actually *is* a common subset that is large enough to be useful. In this approach, if *one* platform doesn't provide a "star" icon, then we wouldn't provide a "star" icon as part of Toga's set.

The second approach is the same as the first, but use a "highest common factor" approach - that is, work out a spanning set of icons that we *could* wrap, and when one platform provides that icon but another doesn't, backfill that missing icon with a Toga-provided one. For this, we'd need to find a source of good open-source compatible, platform-specific icons. If *one* of the platforms provides a "star" icon, then we backfill all the other platforms with a version that fits that platform's style guide. 

The third approach would be to ignore platform icons entirely, and provide integration with a third-party icon set, like [Icons8](https://icons8.com) or [FontAwesome](https://fontawesome.com). There's 2 issues with this approach: licensing (good icon sets usually have commercial terms), and distribution size (why ship 1000 icons when your app only uses 3?). To do this, I suspect we'd need to ship "toga-icons8" (for example) as a standalone package, and then provide additional tooling to make it easy to install specific icons from the complete set, rather than ship *all* icons ready-to-use (or maybe there's a  on-demand approach 

We could possibly support multiple approaches - say, option 2 *and* 3, where we provide a healthy set of builtin icons, and an expansion path for other icon sets. We could even support *multiple* icon sets - either with different iconset-specfic packages, or by using a single "toga-iconset" package that can be configured to use different icons sets by configuration or plugin.

tl;dr - I agree this would be a useful feature; the next step will be for someone to research what icons are available by default, and what options exist for platform-specific icon sets - especially if they're platform specific, and/or freely licensed.

**Comment by johnzhou721** [comment]
Some preliminary research.

iOS, Cocoa: https://developer.apple.com/design/human-interface-guidelines/icons#Standard-icons

Android: https://fonts.google.com/icons (go from https://m3.material.io/styles/icons/overview) -- however, online sources seems to suggest that you must generate pngs in Android studio for it yourself.  I dug around and found https://github.com/mikepenz/Android-Iconics but I'm not sure what set of icons it bundles in the material directory;  alternatively if unofficial but visually compatible icons are fine, there's https://pictogrammers.com/library/mdi/ with https://github.com/code-mc/material-icon-lib/tree/master to use them at runtime with a simpler API.

Winforms: https://grantwinney.com/how-to-use-getstockicon-for-winforms/ (unofficial source), available list at https://learn.microsoft.com/en-us/dotnet/api/system.drawing.stockiconid?view=windowsdesktop-9.0

GTK: Maybe https://docs.gtk.org/gtk4/class.IconTheme.html?  But then we'd need to be able to browse for the icons we can get -- https://askubuntu.com/questions/695796/view-list-of-all-available-unique-icons-with-their-names-and-thumbnail seems to be a pretty good reference for browsing icons.

Qt:  https://www.pythonguis.com/faq/built-in-qicons-pyqt/ -- Qt ships a small set and then you can get more from a freedesktop theme.  I highly suspect the reference mentioned in the previous paragraph will work for browsing available Linux-native QIcon.fromTheme options as well.  I'm mentioning Qt here for another reason -- Qt's shipping of a small set of icons might be able to be used to fill in some unavailable icons on other platforms.

But Qt is LGPLv3, so not sure how those will integrate well into Toga. Since you can link to LGPLv3 without any impediments, perhaps we should make another "toga-qt-fallbacks" package distributing those extra icons along with some simple utils to get them, and release that under LGPLv3, that way we can keep Toga under BSD.

I'm +0 on option 3 because those icons wouldn't look native.

**Comment by freakboy3742** [comment]
> Some preliminary research.

Thanks for that - the next step will be working out:

1. A meaningful initial subset of icons, and a discussion about whether we're going to follow a "lowest common denominator" approach, or a "allow everything and provide fallbacks" approach.
2. An API for Icon that allows using those icons (my initial guess would be `Icon(name=...)` analogous to how `Image` can accept `path` or `data`, but I'm open to other suggestions)

> But Qt is LGPLv3, so not sure how those will integrate well into Toga.

If we can access the Qt icons using a PySide6 API, there's no problem at all. We're not including anything LGPL in our code; we're just using a published API.

If we need to copy the resources from the Qt source tree, then it's more of a problem because we'd need to include those icons under license.

**Comment by johnzhou721** [comment]
Some remarks here:

> > But Qt is LGPLv3, so not sure how those will integrate well into Toga.
> 
> If we can access the Qt icons using a PySide6 API, there's no problem at all. We're not including anything LGPL in our code; we're just using a published API.
> 
> If we need to copy the resources from the Qt source tree, then it's more of a problem because we'd need to include those icons under license.

Probably not with PySide6.  We need a QApplication instance to get icons to initialize properly, and I'm not so sure how we're going to juggle both a QApplication and say NSApplication in one app.

https://github.com/mikepenz/Android-Iconics/ uses the older Material Icons library.

EDIT CLARIFYING -- Cocoa and iOS API is https://developer.apple.com/documentation/uikit/uiimage/init(systemname:)?language=objc https://developer.apple.com/documentation/appkit/nsimage/init(systemsymbolname:accessibilitydescription:)?language=objc

<details>

<summary> List of corresponding icons that are implemented in >=2 of these sets </summary>

| Purpose | Winf. StockIcons | freedesk. ID | SF Sym | Mtl. Icns |
| --------- | -------- | -------  | ------- | ------- |
| New File | DocumentNoAssociation | document-new | document.badge.plus | insert_drive_file |
| Empty / Unknown Document | DocumentNoAssociation | ??? | document | insert_drive_file |
| Document w/ Text | DocumentWithAssociation | accessories-text-editor | text.document | description |
| Application / Executable | Application | application-x-executable | app.grid | apps |
| Closed Folder | Folder | folder | folder | folder |
| Open Folder | FolderOpen | folder-open | folder (no open variant) | folder_open |
| Removable / External Drive/Storage | DriveRemovable | drive-removable-media | externaldrive | sd_storage (too specific?) |
| Fixed / Internal Drive/Storage | DriveFixed | drive-harddisk | internaldrive | storage |
| Network Drive | DriveNet | network-server | externaldrive.connected.to.line.below | folder_shared (?) |
| CD Storage | DriveCD | drive-optical | opticaldiscdrive (not graphically disc, but right semantically) | ??? (perhaps crop from "disc_full") |
| Memory Chip | DriveRam | ??? | memorychip | memory |
| Entire Network | World | preferences-system-network | network | public |
| Printer | Printer | printer | printer | print |
| Find / Search | Find | edit-find | magnifyingglass | search |
| Help | Help | help-browser | questionmark.circle | help_outline |
| Recycle Bin | Recycler | user-trash | trash (?? delete icon; May need custom design) | delete (accurate visual; restore from trash is based on this symbol) |
| Full Recycle Bin | RecyclerFull | user-trash-full | trash-fill (??? may need another design) | ??? (maybe filled delete?) |
| Album / Media CD | MediaCDAudio | media-optical-cd-audio | ??? | album |
| Lock | Lock | system-lock-screen (??? kinda abuse, for "Lock Screen" in system) | lock | lock |

Sources for filling missing ones:
1. Qt's built-in SP_... icon set.  Extract and release as separate. LGPL package that we declare a dep of.
2. There's https://ionic.io/ionicons for Android and iOS
3. Modifying existing icons and releasing them in separate packages with compatible license (Material and freedesktop only), then declaring them deps of Toga to access them

</details>

---

EDIT 1/2/2026:  I'm updating the table above; I'm going to go throuhg both WinForms and freedesktop IDs to get all possible semantic things.  However, I'm only putting it in the table if there's at least 2 platforms that abstracts it satisfyingly; if there's no abstraction for some of them, I may provide ways we could edit it / fill it in w/ compatible.
